streams-contrib/streams-processor-regex/src/main/java/org/apache/streams/regex/RegexUtils.java

```diff
@@ -62,26 +63,21 @@ private RegexUtils() {}
 
     protected static Map<String, List<Integer>> getMatches(String pattern, String content, int capture) {
         try {
-            Map<String, List<Integer>> matches = Maps.newHashMap();
-            if(content == null) {
-                return matches;
-            }
+            //Since certain regex patterns can be susceptible to catastrophic backtracking
+            //We need to be able to isolate this operation in a separate thread and kill
+            //it if it ends up taking too long
+            RegexRunnable regexRunnable = new RegexRunnable(pattern, content, capture);
 
-            Matcher m = getPattern(pattern).matcher(content);
-            while (m.find()) {
-                String group = capture > 0 ? m.group(capture) : m.group();
-                if (group != null && !group.equals("")) {
-                    List<Integer> indices;
-                    if (matches.containsKey(group)) {
-                        indices = matches.get(group);
-                    } else {
-                        indices = Lists.newArrayList();
-                        matches.put(group, indices);
-                    }
-                    indices.add(m.start());
-                }
+            Thread thread = new Thread(regexRunnable);
```

Comment:
My concern in creating a new thread for each document is that you will be incurring quite a bit of overhead in thread management.  

Have you thought through a ForkJoinPool or ExecutorServices vs a thread for each string.  

Thoughts @smashew @sieverssj?